packages/core/src/storage/namespace.ts

```diff
@@ -0,0 +1,305 @@
+      addContractNamespacesWithSrcs(result, decodeSrc, layout, context, origReferencedContract);
+    }
+  }
+
+  return result;
+}
+
+/**
+ * Gets the contract definition for the given referenced id.
+ */
+function getReferencedContract(context: CompilationContext, referencedId: number) {
+  // Optimization to avoid dereferencing if the referenced id is the same as the current contract
+  return context.contractDef.id === referencedId
+    ? context.contractDef
+    : context.deref(['ContractDefinition'], referencedId);
+}
+
+/**
+ * Add namespaces and source locations for the given compilation context's contract.
+ * Does not include inherited contracts.
+ */
+function addContractNamespacesWithSrcs(
+  namespaces: Record<string, NamespaceWithSrcs>,
+  decodeSrc: SrcDecoder,
+  layout: StorageLayout,
+  context: CompilationContext,
+  origContractDef: ContractDefinition,
+) {
+  for (const node of context.contractDef.nodes) {
+    if (isNodeType('StructDefinition', node)) {
+      const storageLocationArg = getStorageLocationArg(node);
+      if (storageLocationArg !== undefined) {
+        const origSrc = decodeSrc(getOriginalStruct(node.canonicalName, origContractDef));
+
+        if (namespaces[storageLocationArg] !== undefined) {
+          namespaces[storageLocationArg].srcs.push(origSrc);
+        } else {
+          namespaces[storageLocationArg] = {
+            namespace: getNamespacedStorageItems(node, decodeSrc, layout, context, origContractDef),
+            srcs: [origSrc],
+          };
+        }
+      }
+    }
+  }
+}
+
+/**
+ * Gets the storage location string from the `@custom:storage-location` annotation.
+ *
+ * For example, when using ERC-7201 (https://eips.ethereum.org/EIPS/eip-7201), the result will be `erc7201:<NAMESPACE_ID>`
+ *
+ * @param node The node that may have a `@custom:storage-location` annotation.
+ * @returns The storage location string, or undefined if the node does not have a `@custom:storage-location` annotation.
+ */
+export function getStorageLocationArg(node: Node) {
+  const doc = getDocumentation(node);
+  if (hasAnnotationTag(doc, 'storage-location')) {
+    return getStorageLocation(doc);
+  }
+}
+
+function getStorageLocation(doc: string) {
+  const storageLocationArgs = getAnnotationArgs(doc, 'storage-location');
+  if (storageLocationArgs.length !== 1) {
+    throw new Error('@custom:storage-location annotation must have exactly one argument');
+  }
+  const storageLocation = storageLocationArgs[0];
+  return storageLocation;
+}
+
+/**
+ * Gets the storage items for the given struct node.
+ * Includes loading recursive type information, and adds slot and offset if they are available in the given compilation context's layout.
+ */
+function getNamespacedStorageItems(
+  node: StructDefinition,
+  decodeSrc: SrcDecoder,
+  layout: StorageLayout,
+  context: CompilationContext,
+  origContractDef: ContractDefinition,
+) {
```

Comment:
It's not strictly necessary but for these complex functions especially I would declare the return type explicitly:
```suggestion
): StorageItem[] {
```